dms/models/directory.py

```diff
@@ -194,6 +195,38 @@ class DmsDirectory(models.Model):
                 """,
     )
 
+    def check_access_rule(self, operation):
+        super().check_access_rule(operation)
+        for record in self:
+            if (
+                record.is_root_directory
+                and not record.root_storage_id.inherit_access_from_parent_record
+            ) or (
+                not record.is_root_directory
+                and not record.storage_id.inherit_access_from_parent_record
+            ):
+                continue
+            # Check access to inherited model (and record)
+            model = self.env[record.res_model]
+            model.check_access_rights(operation)
+            if record.res_id:
+                model.browse(record.res_id).check_access_rule(operation)
+        records = self - self.filtered(
+            lambda x: x.res_model
+            and (
+                (
+                    x.is_root_directory
+                    and x.root_storage_id.inherit_access_from_parent_record
+                )
+                or (
+                    not x.is_root_directory
+                    and x.storage_id.inherit_access_from_parent_record
+                )
+            )
+        )
+        for record in records:
+            record.check_access_groups(operation)
```

Comment:
Some comments about this code chunk:

1. When `directory.is_root_directory == True`, `storage_id` will be the same as `root_storage_id`: https://github.com/OCA/dms/blob/2fb661a5ef86afdf7eff51e6bdc1a46cb6591a7c/dms/models/directory.py#L381-L382
2. It seems you are entering a loop and discarding some elements, and then generating that loop again with the other elements. It would be easier to just loop everything and do a if-else.

So, this seems to be the same, but shorter:

```suggestion
        for record in self:
            if not record.storage_id.inherit_access_from_parent_record:
                record.check_access_groups(operation)
                continue
            # Check access to inherited model (and record)
            model = self.env[record.res_model]
            model.check_access_rights(operation)
            if record.res_id:
                model.browse(record.res_id).check_access_rule(operation)
```

Reply:
I'm not really sure if these change will have side effects. Something similar is commented here https://github.com/OCA/dms/pull/67#discussion_r608528935
IMO it's a good idea to change root_storage_id to storage_id and add  consistency to the fields BUT it's important to ask to @etobella (probablly kwnow more about it and if we can change it) and apply these change in other PR about to code re-factoring.